### PR TITLE
hypercore v5  support

### PR DESCRIPTION
**Pull request by okdistribute**
started SLEEP (hypercore v5) support, but missing something in the replicate function.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -223,10 +222,11 @@ function create (opts) {
     opened[hex] = old
 
     if (!feed) {
-      old.feed = feed = core.createFeed(key, {
-        storage: storage(path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data')),
-        sparse: opts.sparse && isContent
-      })
+      old.feed = feed = hypercore(
+        storage(path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data')),
```
Comment (mafintosh) [inline index.js]: just pass `path.join(dir, 'data', key.slice(0, 2), key.slice(2))` here instead

Comment (okdistribute) [reply index.js]: ok cool got it

File: index.js [hunk 2]
```diff
@@ -170,15 +170,13 @@ function create (opts) {
...
-      })
+      var feed = hypercore(storage(path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data')), key)
+
       noContent.get(discKey, function (err) {
         if (!err) return done(null)
         feed.get(0, {wait: false}, function (err, data) {
           if (err) {
-            if (err.notFound) return done(null)
-            return cb(err)
+            if (!err.notFound) return cb(err)
```
Comment (mafintosh) [inline index.js]: hmm why this change? (will make it crash it there was an notFound error?)

File: index.js [hunk 3]
```diff
@@ -199,10 +197,11 @@ function create (opts) {
         if (!contentKey) return cb(null, feed)
 
         contentKey = datKeyAs.str(contentKey)
-        var contentFeed = core.createFeed(contentKey, {
-          storage: storage(path.join(dir, 'data', contentKey.slice(0, 2), contentKey.slice(2) + '.data')),
-          sparse: opts.sparse
-        })
+        var contentFeed = hypercore(
+          storage(path.join(dir, 'data', contentKey.slice(0, 2), contentKey.slice(2) + '.data')),
```
Comment (mafintosh) [inline index.js]: same as below

File: index.js [hunk 4]
```diff
@@ -170,15 +170,13 @@ function create (opts) {
       if (err) return cb(err) // no key found
 
       key = datKeyAs.str(key)
-      var feed = core.createFeed(key, {
-        storage: storage(path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data'))
-      })
+      var feed = hypercore(storage(path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data')), key)
```
Comment (mafintosh) [inline index.js]: same as below

File: index.js [hunk 5]
```diff
@@ -38,9 +38,9 @@ function create (opts) {
 
   that.changes = thunky(function (cb) {
     misc.get('changes', function (_, key) {
-      var feed = core.createFeed(key)
+      var feed = hypercore(storage, key)
 
-      feed.open(function (err) {
+      feed.on('open', function (err) {
```
Comment (mafintosh) [inline index.js]: the event is called 'ready'

File: index.js (hunk 2 above)
Comment (okdistribute) [reply index.js]: not sure how that got in there, I think I was experimenting on another branch. removing it

File: index.js (hunk 5 above)
Comment (okdistribute) [reply index.js]: got it

File: index.js [hunk 6]
```diff
@@ -23,7 +23,7 @@ function create (opts) {
   var misc = subleveldown(db, 'misc', {valueEncoding: 'binary'})
   var keys = subleveldown(db, 'added-keys', {valueEncoding: 'binary'})
   var noContent = subleveldown(db, 'no-content', {valueEncoding: 'binary'})
-  var core = hypercore(db)
+  var core = hypercore(storage)
```
Comment (mafintosh) [inline index.js]: this line should just be removed

File: index.js (hunk 2 above)
Comment (mafintosh) [reply index.js]: 👍

File: index.js (hunk 6 above)
Comment (okdistribute) [reply index.js]: replicate() depends upon the `core` object, does that mean the replicate function will need to return a generic stream and pipe through the requested hypercore keys?

Comment (mafintosh) [reply index.js]: for this use cause yes. var stream = protocol() ... feed.replicate({stream: stream})

File: index.js [hunk 7]
```diff
@@ -223,14 +222,15 @@ function create (opts) {
...
+      old.feed = feed = hypercore(
+        path.join(dir, 'data', key.slice(0, 2), key.slice(2) + '.data'),
+        key,
+        {sparse: opts.sparse && isContent}
+      )
     }
 
     old.cnt++
-    feed.replicate({stream: stream})
+    var replicator = feed.replicate({stream: stream})
```
Comment (okdistribute) [inline index.js]: @mafintosh not sure if my logic is right here on replication, seems like replicator.on('end' is never firing

Comment (okdistribute) [reply index.js]: ah so stream.on('open') is never firing above

Comment (mafintosh) [reply index.js]: its called feed now, https://github.com/mafintosh/hypercore-protocol#streamonfeed-discoverykey

Comment (okdistribute) [reply index.js]: nice. ok so still replicator.end is never firing

File: index.js [hunk 8]
```diff
@@ -223,23 +224,27 @@ function create (opts) {
...
     if (!feed.firstDownload) {
       var downloaded = false
 
       feed.firstDownload = true
       feed.once('download', function () {
+        console.log('downloaded')
         downloaded = true
       })
-      feed.once('download-finished', function () {
+      replicator.on('end', function () {
```
Comment (okdistribute) [inline index.js]: This is never getting called, is there a better way to check if the feed has been downloaded?

**Comment by okdistribute** [comment]
close, but `tests/archive.js` is still failing now with this:

```
/Users/karissa/node_modules/hypercore-archiver/node_modules/hyperdrive-encoding/index.js:24
  if (!enc) throw new Error('Unknown message type: ' + toTypeString(type))
            ^

Error: Unknown message type: unknown
    at getEncoding (/Users/karissa/node_modules/hypercore-archiver/node_modules/hyperdrive-encoding/index.js:24:19)
    at Object.decode (/Users/karissa/node_modules/hypercore-archiver/node_modules/hyperdrive-encoding/index.js:32:13)
    at /Users/karissa/node_modules/hypercore-archiver/tests/archive.js:52:28
    at onread (/Users/karissa/node_modules/hypercore-archiver/node_modules/random-access-file/index.js:107:25)
    at FSReqWrap.wrapper [as oncomplete] (fs.js:681:17)
npm ERR! Test failed.  See above for more details.
```

**Comment by okdistribute** [comment]
and `test/feed.js` with 
```
/Users/karissa/node_modules/hypercore-archiver/node_modules/hypercore/lib/replicate.js:269
  this.stream.want({start: 0}) // TODO: don't just subscribe to *EVERYTHING* hehe
             ^

TypeError: Cannot read property 'want' of null
    at Peer.ready (/Users/karissa/node_modules/hypercore-archiver/node_modules/hypercore/lib/replicate.js:269:14)
    at /Users/karissa/node_modules/hypercore-archiver/node_modules/hypercore/lib/replicate.js:24:10
    at /Users/karissa/node_modules/hypercore-archiver/node_modules/hypercore/index.js:113:15
    at apply (/Users/karissa/node_modules/hypercore-archiver/node_modules/thunky/index.js:44:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by joehand** [comment]
@karissa I think I fixed the content issue, we had to change how we were doing the decoding of the index block. But... something is still broken with the decoding. I think we need to catch an error that we aren't right now but can't quite figure it out where the test error is coming from:

```
# add new feed key
ok 28 changes error
ok 29 changes feed get error
ok 30 one block in change feed
ok 31 changes key okay
ok 32 changes type add
ok 33 add event key okay
ok 34 add error
ok 35 archived event key okay
ok 36 feed get error
ok 37 feed block: 0 is correct
/Users/joe/node_modules/hypercore-archiver/node_modules/protocol-buffers/compile.js:32
      throw new Error('Groups are not supported')
      ^

Error: Groups are not supported
    at skip (/Users/joe/node_modules/hypercore-archiver/node_modules/protocol-buffers/compile.js:32:13)
    at Object.decode (eval at <anonymous> (/Users/joe/node_modules/hypercore-archiver/node_modules/generate-function/index.js:55:21), <anonymous>:33:16)
    at /Users/joe/node_modules/hypercore-archiver/node_modules/hypercore/index.js:967:18
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
npm ERR! Test failed.  See above for more details.
```

That's about all my brain can handle for the day!

**Comment by okdistribute** [comment]
Going to take a stab at it. Found a few reference issues

https://github.com/mafintosh/protocol-buffers/issues/38
https://github.com/mafintosh/protocol-buffers/issues/44

**Comment by okdistribute** [comment]
If its any consolation, it works! :P (just the tests are broken)

**Comment by okdistribute** [comment]
OK I added some testing to the tests/archive.js -- I realize that if you keep the new valueEncoding stuff in `index.js`, the new archive tests pass, but if you remove valueEncoding, these new archive tests fail.

So the Group bug comes when you're trying to read a hypercore as though its a hyperdrive, because protobuf is trying to decode a string using message.Index 

So we need some way to specify (or see inside) if a feed is a hyperdrive and encode the values accordingly if so.

**Comment by okdistribute** [comment]
I think I got it in there, tell me what you think

**Comment by clkao** [comment]
We should bump version to v4?

**Comment by mafintosh** [comment]
@karissa i investigated your bug in hypercore today. it is a bug on my side but it is caused by the fact that the *first* hypercore shared is *not* the same which it should be. you prob just need to guard some async logic to make sure that happens. ping me if this makes no sense :)

**Comment by joehand** [comment]
@karissa updated the test. I was having trouble with the old test and the feeds weren't ever added to the archiver.

This one is a bit more explicit about the adding + replication. BUT still having trouble replicating the second feed.

However, it doesn't look like it's related to the order - they are getting opened and replicated in the correct order. Still digging in...

**Comment by okdistribute** [comment]
OK Yeah, that test looks better, nice!

**Comment by clkao** [comment]
Here's the change to allow archiver.get to optionally use timeout.

```
commit 8b1059478b31c8e21fe6eb3b821458d8e54c7e68
Author: Chia-liang Kao <clkao@clkao.org>
Date:   Thu Apr 20 14:24:29 2017 +0800

    allow feed.get options

diff --git a/index.js b/index.js
index fd925ae..6ad17b2 100644
--- a/index.js
+++ b/index.js
@@ -165,7 +165,8 @@ function create (opts) {
     })
   }
 
-  function get (key, cb) {
+  function get (key, opts, cb) {
+    if (typeof opts === 'function') return get(key, {wait: false}, opts)
     key = datKeyAs.buf(key)
     var discKey = hypercore.discoveryKey(key).toString('hex')
     keys.get(discKey, function (err, key) {
@@ -176,18 +177,14 @@ function create (opts) {
 
       noContent.get(discKey, function (err) {
         if (!err) return done(null)
-        feed.get(0, {
-          wait: false
-        }, function (err, data) {
+        feed.get(0, opts, function (err, data) {
           if (err) {
             if (err.notFound) return done(null)
             return cb(err)
           }
           var decoded = tryHyperdrive(data)
           if (decoded.content || !feed.length) return done(decoded.content)
-          feed.get(feed.length - 1, {
-            wait: false
-          }, function (err, data) {
+          feed.get(feed.length - 1, opts, function (err, data) {
             if (err) {
               if (err.notFound) return done(null)
               return cb(err)

```

File: tests/feed.js [hunk 9]
```diff
@@ -95,7 +95,7 @@ test('replicate to hypercore from archiver', function (t) {
 })
 
 test('replicate two hypercores from archiver', function (t) {
-  t.plan(6)
+  t.plan(4)
```
Comment (joehand) [inline tests/feed.js]: Ah ya, 6 is correct because `thisArchives.on('archived')` should emit twice, once for each feed. 

This is why I think the double append is still buggy. The second feed never emits archived. Definitely confused about this bit so please double check my logic =).

**Comment by mafintosh** [comment]
Rewrote the entire thing. Latest npm release (4) has v5 support
